python/adjoint/filters.py

```diff
@@ -694,9 +695,164 @@ def tanh_projection(x: np.ndarray, beta: float, eta: float) -> np.ndarray:
+    # and determine where smoothing is actually needed. The value where we don't
+    # need smoothings doesn't actually matter, since all our computations our
+    # purely element-wise (no spatial locality) and those pixels will instead
+    # rely on the standard projection.
+    nonzero_norm = npa.abs(x_grad_helper) > sys.float_info.epsilon
+
+    x_grad_norm = npa.sqrt(npa.where(nonzero_norm, x_grad_helper, 1))
+    x_grad_norm_eff = npa.where(nonzero_norm, x_grad_norm, 1)
+
+    # The distance for the center of the pixel to the nearest interface
+    d = (eta - x_smoothed) / x_grad_norm_eff
+
+    # Only need smoothing if an interface lies within the voxel. Since d is
+    # actually an "effective" d by this point, we need to ignore values that may
+    # have been sanitized earlier on.
+    needs_smoothing = nonzero_norm & (npa.abs(d) <= pixel_radius)
+    d = npa.where(needs_smoothing, d, 1)
+
+    # The fill factor is used to perform simple, first-order subpixel smoothing.
+    # We use the (2D) analytic expression that comes when assuming the smoothing
+    # kernel is a circle. Note that because the kernel contains some
+    # expressions that are sensitive to NaNs, we have to use the "double where"
+    # trick to avoid the Nans in the backward trace. This is a common problem
+    # with array-based AD tracers, apparently. See here:
+    # https://github.com/google/jax/issues/1052#issuecomment-5140833520
+
+    arccos_term = pixel_radius**2 * npa.arccos(
+        npa.where(
+            needs_smoothing,
+            d / pixel_radius,
+            0.0,
+        )
+    )
 
-    return (npa.tanh(beta * eta) + npa.tanh(beta * (x - eta))) / (
-        npa.tanh(beta * eta) + npa.tanh(beta * (1 - eta))
+    sqrt_term = d * npa.sqrt(
+        npa.where(
+            needs_smoothing,
+            pixel_radius**2 - d**2,
+            1,
+        )
+    )
+    fill_factor = (1 / (npa.pi * pixel_radius**2)) * (arccos_term - sqrt_term)
+    fill_factor_eff = npa.where(
+        needs_smoothing,
+        fill_factor,
+        1,
+    )
+
+    # Determine the upper and lower bounds of materials in the current pixel.
+    x_minus = x_smoothed - x_grad_norm_eff * pixel_radius
+    x_plus = x_smoothed + x_grad_norm_eff * pixel_radius
+
+    # Create an "effective" set of materials that will ensure everything is
+    # piecewise differentiable, even if an interface moves out of a pixel, or
+    # through the pixel center.
+    x_minus_eff_pert = (x_smoothed * d + x_minus * (pixel_radius - d)) / pixel_radius
+    x_minus_eff = npa.where(
+        (d > 0),
+        x_minus_eff_pert,
+        x_minus,
+    )
+    x_plus_eff_pert = (-x_smoothed * d + x_plus * (pixel_radius + d)) / pixel_radius
+    x_plus_eff = npa.where(
+        (d > 0),
+        x_plus,
+        x_plus_eff_pert,
+    )
+
+    # Only apply smoothing to interfaces
+    x_projected_smoothed = (1 - fill_factor_eff) * x_minus_eff + (
+        fill_factor_eff
+    ) * x_plus_eff
```

Comment:
> Check smoothness numerically.

Hmm so after running @stevengj's experiment, we can clearly see the discontinuity at play here:

![image](https://github.com/NanoComp/meep/assets/24902086/d84344de-1be9-4eb8-a7cf-bbcb6d98f825)

To be precise, we still get the continuous "sigmoid-like" transition (from the 2d case of $F_c$), but it's not being scaled properly. So it can't transition properly from/to 0/1.

The code above shows how this scaling is performed.

Here's what happens if we use `beta=8`:

![image](https://github.com/NanoComp/meep/assets/24902086/283c3e23-5d89-45ec-814c-af20505ef070)

Reply:
Ah Ok I think I've figured it out. Essentially, I was forgetting to _project_ our endpoints too (which affects the scaling). When I do this, We get the following results:

`beta=np.inf`
![image](https://github.com/NanoComp/meep/assets/24902086/7949cacb-ef4f-4c27-8b7b-6c867b808da4)

`beta=8`
![image](https://github.com/NanoComp/meep/assets/24902086/fb972aba-7367-4315-9697-7bee1f1ebfe0)

(note that the derivative is still not continuous in these cases. But that's somewhat expected, since we don't take the derivative at the edges into account when constructing $F_c$. So a more "intelligent" choice for $F_c$ here might be something that looks more like a Cubic hermite spline or something.)

The gradient convergence still looks good:
<img width="511" alt="image" src="https://github.com/NanoComp/meep/assets/24902086/3a6897a2-83e7-4f5b-94af-5dc44fbafb51">


And the convergence plot no longer stalls:
![Figure_1](https://github.com/NanoComp/meep/assets/24902086/423da3e7-7ad2-411e-a3d9-932dc0855763)

Reply:
Will this affect https://github.com/NanoComp/photonics-opt-testbed/pull/63, since that used a Julia port of this code?

Reply:
> since that used a Julia port of this code?

Yep I'm talking with @hammy4815 offline and he's trying out an updated version to see if this resolves some of the challenges he was seeing.